pkg/remote/remotecli/client_test.go

```diff
@@ -0,0 +1,310 @@
+		WriteFunc: func(ctx context.Context, conn net.Conn, send remote.Message) error {
+			isWrite = true
+			writeMsg = send
+			return errors.New(errMsg)
+		},
+		ReadFunc: nil,
+	}
+
+	opt := newMockOption(addr)
+
+	isClose := false
+	conn := &mocks.Conn{
+		RemoteAddrFunc: func() (r net.Addr) {
+			return addr
+		},
+		CloseFunc: func() (err error) {
+			isClose = true
+			return err
+		},
+	}
+	opt.Dialer = &remote.SynthesizedDialer{
+		DialFunc: func(network, address string, timeout time.Duration) (net.Conn, error) {
+			return conn, nil
+		},
+	}
+
+	// 2. test
+	cli, err := NewClient(ctx, ri, hdlr, opt)
+	test.Assert(t, err == nil, err)
+	test.Assert(t, cli != nil, cli)
+
+	var sendMsg remote.Message
+	err = cli.Send(ctx, ri, sendMsg)
+	test.Assert(t, err != nil, err)
+	test.Assert(t, err.Error() == errMsg)
+	test.Assert(t, isWrite)
+	test.Assert(t, writeMsg == sendMsg)
+	test.Assert(t, isClose)
+}
+
+// TestRecv test recv msg
+func TestRecv(t *testing.T) {
+	ctx := context.Background()
+
+	addr := utils.NewNetAddr("tcp", "to")
+	ri := newMockRpcInfo(addr)
+
+	svcInfo := mocks.ServiceInfo()
+	var resp interface{}
+	msg := remote.NewMessage(resp, svcInfo, ri, remote.Call, remote.Client)
+
+	var readMsg remote.Message
+	isRead := false
+	hdlr := &mocks.MockCliTransHandler{
+		WriteFunc: nil,
+		ReadFunc: func(ctx context.Context, conn net.Conn, msg remote.Message) error {
+			isRead = true
+			readMsg = msg
+			return nil
+		},
+	}
+
+	opt := newMockOption(addr)
+
+	cli, err := NewClient(ctx, ri, hdlr, opt)
+	test.Assert(t, err == nil, err)
+	test.Assert(t, cli != nil, cli)
+
+	err = cli.Recv(ctx, ri, msg)
+	test.Assert(t, err == nil, err)
+	test.Assert(t, isRead)
+	test.Assert(t, readMsg == msg)
+}
+
+// TestRecvOneWay test recv onw way msg
+func TestRecvOneWay(t *testing.T) {
+	ctx := context.Background()
+
+	addr := utils.NewNetAddr("tcp", "to")
+	ri := newMockRpcInfo(addr)
+
+	hdlr := &mocks.MockCliTransHandler{
+		WriteFunc: nil,
+		ReadFunc:  nil,
+	}
+
+	opt := newMockOption(addr)
+
+	isClose := false
+	conn := &mocks.Conn{
+		RemoteAddrFunc: func() (r net.Addr) {
+			isClose = true
+			return addr
+		},
+		CloseFunc: func() (err error) {
+			return err
+		},
+	}
+	opt.Dialer = &remote.SynthesizedDialer{
+		DialFunc: func(network, address string, timeout time.Duration) (net.Conn, error) {
+			return conn, nil
+		},
+	}
+
+	cli, err := NewClient(ctx, ri, hdlr, opt)
+	test.Assert(t, err == nil, err)
+	test.Assert(t, cli != nil, cli)
+
+	err = cli.Recv(ctx, ri, nil)
+	test.Assert(t, err == nil, err)
+	test.Assert(t, isClose)
+}
+
+// TestRecycle test recycle
+func TestRecycle(t *testing.T) {
+	ctx := context.Background()
+
+	addr := utils.NewNetAddr("tcp", "to")
+	ri := newMockRpcInfo(addr)
+
+	hdlr := &mocks.MockCliTransHandler{
+		WriteFunc: nil,
+		ReadFunc:  nil,
+	}
+
+	opt := newMockOption(addr)
+
+	cli, err := NewClient(ctx, ri, hdlr, opt)
+	test.Assert(t, err == nil, err)
+	test.Assert(t, cli != nil, cli)
+
+	cli.Recycle()
```

Comment:
check fields of cli again after recycle

Reply:
It has been modified，because cli's attr is private variable，but Recycle function has put cli into clientPool，so I check  cli is equal